### PR TITLE
New textEditorDialog

### PR DESCRIPTION
**Pull request by Bouh**

### CODE REVIEW

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+  sizeTextField: {
+    width: 90,
+    ...toolbarItemStyle,
+  },
+  toolbarItem: toolbarItemStyle,
+  checkbox: toolbarItemStyle,
+};
+
+type Props = {|
+  event: any,
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: event: gdBaseEvent

(even if it's currently almost equivalent to any, in the future GDevelop.js could autogenerate these types)

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+  onClose: () => void,
+  onApply: () => void,
+|};
+
+type State = {|
+  textValue: any,
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: put proper types there :) This will save us from silly mistakes.

textValue is a `string`
textColors and backgroundColors could be objects like this: `{|r:number, g: number, b: number|}`

(the burger notation `{| |}` indicates that the object can not have other properties than the one listed)

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    }
+  };
+
+  setComment = () => {
+    const { event } = this.props;
+    const eventType = event.getType();
+    if (eventType === 'BuiltinCommonInstructions::Comment') {
+      gd.asCommentEvent(event).setComment(this.state.textValue);
+    }
+  };
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: Could you group these two methods in a single one called `_applyChangesOnEvent`?

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        parseInt(backgroundColors.blue)
+      );
+    } else if (eventType === 'BuiltinCommonInstructions::Group') {
+      gd.asGroupEvent(event).setBackgroundColor(
+        parseInt(backgroundColors.red),
+        parseInt(backgroundColors.green),
+        parseInt(backgroundColors.blue)
+      );
+    }
+  };
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: I think we could also make it part of `_applyChangesToEvent`

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+      };
+
+      return backgroundColors;
+    } else {
+      console.error(
+        'Dialog was opened for an unsupported event background color'
+      );
+      return '';
+    }
+  };
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: And these 3 methods could be a single `_getInitialStateFromEvent`.
I think this would reduce some boilerplate and the number of if/else to do (once instead of 3 times).

The object you should return is the `State`: `state = this._getInitialStateFromEvent()`

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    backgroundColors: this.getBackgroundColors(),
+  };
+
+  render() {
+    const { event, onApply, onClose } = this.props;
+    const eventType = event.getType();
+
+    return (
+      <Dialog
+        title={<Trans>Text editor</Trans>}
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: A bit skeptical about this, the text editor might be confused with the editor of a text object... maybe "Edit the event text"?

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+            primary
+            keyboardFocused
+            onClick={() => {
+              if (eventType === 'BuiltinCommonInstructions::Comment') {
+                this.setComment();
+              } else if (eventType === 'BuiltinCommonInstructions::Group') {
+                this.setName();
+              }
+              this.setTextColors();
+              this.setBackgroundColors();
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: See my previous comment, ideally we do this as a single operation in a function, could make things a bit clearer.

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+            </MiniToolbarText>
+            <ColorPicker
+              style={styles.sizeTextField}
+              disableAlpha
+              color={{
+                r: this.state.backgroundColors.red,
+                g: this.state.backgroundColors.green,
+                b: this.state.backgroundColors.blue,
+                a: 255,
+              }}
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: Ideally, we could store in the state objects with the same shape as the ones passed to the ColorPicker:

colors={this.state.backgroundColor}

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+              onChangeComplete={color => {
+                this.setState({
+                  backgroundColors: {
+                    red: color.rgb.r,
+                    green: color.rgb.g,
+                    blue: color.rgb.b,
+                  },
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: `...and here we could do this.setState({ backgroudColor: color })`

if we use the same shape for all colors.

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+                });
+
+                this.forceUpdate();
+              }}
+            />
+
+            {eventType !== 'BuiltinCommonInstructions::Group' && (
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: Instead of excluding an event type, prefer to test for the event type that is the one that need the components (i.e: the comment).
Otherwise if for some reason a bad event is passed here, the component will be shown and could make things worse for the user.

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+                    a: 255,
+                  }}
+                  onChangeComplete={color => {
+                    this.setState({
+                      textColors: {
+                        red: color.rgb.r,
+                        green: color.rgb.g,
+                        blue: color.rgb.b,
+                      },
+                    });
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: Same remarks relative to simplifying color objects

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+                    this.forceUpdate();
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: forceUpdate is not needed after a setState

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+          <Line noMargin>
+            <Column expand>
+              <Line>
+                <SemiControlledTextField
+                  commitOnBlur
+                  hintText={t`Enter the text to be displayed`}
+                  fullWidth
+                  multiLine
+                  rows={100}
+                  rowsMax={100}
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: That's too much! :) Let's use 8 as in the text editor.

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+                  value={this.state.textValue}
+                  onChange={value => {
+                    this.setState({
+                      textValue: value,
+                    });
+                    this.forceUpdate();
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: no forceUpdate

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+                  hintText={t`Enter the text to be displayed`}
+                  fullWidth
+                  multiLine
+                  rows={100}
+                  rowsMax={100}
+                  value={this.state.textValue}
+                  onChange={value => {
+                    this.setState({
+                      textValue: value,
+                    });
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: nitPicking: 
```
onChange={textValue =>
                    this.setState({
                      textValue
                    })
```

File: newIDE/app/src/EventsSheet/index.js [hunk 16]
```diff
@@ -133,6 +134,8 @@ type State = {|
 
   serializedEventsToExtract: ?Object,
 
+  textEditedEvent: ?any,
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/index.js]: ?gdBaseEvent

File: newIDE/app/src/UI/TextField.js [hunk 17]
```diff
@@ -75,7 +75,7 @@ type Props = {|
   margin?: 'none' | 'normal',
   fullWidth?: boolean,
   style?: {|
-    fontSize?: 18,
+    fontSize?: number | string,
```
Comment (4ian) [inline newIDE/app/src/UI/TextField.js]: Might not be needed in this PR?

**Comment by Wend1go** [comment]
Do you intend to add buttons for setting the outline and text shadow to the text editor dialog?

**Comment by 4ian** [comment]
@Wend1go this is for editing the color and text of comments and group events ;)

@Bouh I think Wend1go comment is revealing that the name of the dialog is not intuitive. This should be related to events, something like "EventTextDialog".

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js (hunk 5 above)
Comment (Bouh) [reply newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: I didn't understand return here

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js (hunk 2 above)
Comment (Bouh) [reply newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: I put "any" by default everywhere except when I know the type ;)
I was not sure if an object was a string or other type, thx !

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js (hunk 5 above)
Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: This return (`return '';`) is incorrect: it's returning a string, while the rest of the function is returning an object. You should then return an object with the same shape (same "type") as the rest of the function.

In other words, surely something like this: `return {r: 0, g: 0, b: 0}; // Return black in case of error.` - though in theory this should never ever be used (in practice, better make the code robust again something passing a wrong event type to this dialog).

**Comment by Bouh** [comment]
> Do you intend to add buttons for setting the outline and text shadow to the text editor dialog?

Not in this PR, this is on my todo list but I've too much todo before, if you give a try before me :) 

> I think Wend1go comment is revealing that the name of the dialog is not intuitive. This should be related to events, something like "EventTextDialog".

Ahah yes i will rename it !


I need your feddback for **this.state.backgroundColor**, I'am lost inside type of colors for set state.
And i can't type directly RGBColor ?

File: newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js (hunk 13 above)
Comment (Bouh) [reply newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: I set 100 for having a large window, because if you put more text in the window, textfield doesn't get bigger because limited by 8 rows.
We need to reopen the editor.

There is also a problem with <minitoolbar> and <dialog>.
Minitoolbar should not be included in the scroll of the dialog.
Maybe we need to have a new props "header" for the dialog. And we pass to header the minitoolbar. 

This is also anoying in for "edit hitboxes" "edit points" "add an animation+"

Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/TextEditorDialog.js]: > Minitoolbar should not be included in the scroll of the dialog.
> Maybe we need to have a new props "header" for the dialog. And we pass to header the minitoolbar.

Yes, I think this could maybe be solved on a case by case basis with the `<ScrollView>` component too (I'm a bit worried of modifying Dialog, I like to keep it simple) (though it's always possible to also make a component DialogWithHeader that uses ScrollView for the middle part). There is something to do.

> I set 100 for having a large window, because if you put more text in the window, textfield doesn't get bigger because limited by 8 rows.
We need to reopen the editor.

Can you set rows={a small number} and rowsMax={a higher number, like 30}? So that the field is taking more space but not too much.

**Comment by Bouh** [comment]
Well i didn't understand what wrong, it's like state are not usable in render().
But the value are define at the end of _getInitialStateFromEvent().
And the render say your state is null ...

**Comment by Bouh** [comment]
Omg forget it, I found why >.>

**Comment by Bouh** [comment]
Is working, but i'm unable to fix errors from flow...

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+  },
+  toolbarItem: toolbarItemStyle,
+  checkbox: toolbarItemStyle,
+};
+type RGBColor = {|
+  r?: number,
+  g?: number,
+  b?: number,
+  a?: number,
+|};
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Not too bad, but you probably don't want to mark r/g/b/a as optional. When doing this, then this means that you consider that these properties can be missing. It's a pain because that mean that you have to check everywhere if r/g/b and a are not undefined before using them! Instead, just say that a RGBColor is *always* containing these variables:

```
type RGBColor = {|
  r: number,
  g: number,
  b: number,
  a: number,
|};
```

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+
+      //Text color for group not supported in Core, instead GroupEvent.js handle this
+      //Background color
+      gd.asGroupEvent(event).setBackgroundColor(
+        parseInt(backgroundColor.r),
+        parseInt(backgroundColor.g),
+        parseInt(backgroundColor.b)
+      );
+      return;
+    }
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Just by removing the optional properties in RGBColor type (see my previous comment), this should solve all the flow errors that you had in this function.

Why? Because you are using the r/g/b properties directly, so if these properties are marked as optional, Flow is warning you: "You're passing a varaible that is maybe undefined to parseInt, which expect a number! Beware there is a bug hidden here!!".

If properties are not marked as optional, Flow will understand that you're passing something that is guaranteed to be a number to parseInt - si it's fine.

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+    const { textValue, textColor, backgroundColor } = this.state;
+    const eventType = event.getType();
+
+    if (eventType === 'BuiltinCommonInstructions::Comment') {
+      //Text value
+      gd.asCommentEvent(event).setComment(textValue);
+
+      //Text color
+      gd.asCommentEvent(event).setTextColor(
+        parseInt(textColor.r),
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Always always always specify the second argument to parseInt: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

```suggestion
        parseInt(textColor.r, 10),
```

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+      gd.asGroupEvent(event).setBackgroundColor(
+        parseInt(backgroundColor.r),
+        parseInt(backgroundColor.g),
+        parseInt(backgroundColor.b)
+      );
+      return;
+    }
+  };
+
+  _getInitialStateFromEvent = () => {
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: To help you, define the type of what your function is returning:
```suggestion
  _getInitialStateFromEvent = (): State => {
```

This will limit errors and the amount of guessing that Flow has to do, making easier to understand error messages.

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+    const { event } = this.props;
+    const eventType = event.getType();
+    let text: string, textColors: RGBColor, backgroundColors: RGBColor;
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: This is kind of "false" as you're declaring variables to be of a certain type, but they are actually undefined :) 
So either you mark them as optional (but don't do that) or (better) you initialize them to default values:

```
let text: string = "";
let textColors: RGBColor = {r:0, g:0,b:0, a:0};
let backgroundColors: RGBColor = {r:0, g:0,b:0, a:0};
```

Otherwise, Flow will warn you: you might be using a variable that is uninitialized! (i.e.: that is undefined).
image

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+      console.error(
+        'Dialog was opened for an unsupported event type: ' + eventType
+      );
+    }
+
+    return {
+      textValue: text,
+      textColor: textColors,
+      backgroundColor: backgroundColors,
+    };
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Rule of thumb: if you assign "something" to "something else", they represent the same thing. So give them the same name :) 

```
     return {
      textValue: textValue,
      textColor: textColor,
      backgroundColor: backgroundColor,
    };
```

which can be simplified to:

```
     return {
      textValue,
      textColor,
      backgroundColor,
    };
```

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js (hunk 22 above)
Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: See my next comment about rename variables too to give them the same name as in the State.

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+        g: groupEvent.getBackgroundColorG(),
+        b: groupEvent.getBackgroundColorB(),
+      };
+
+      text = gd.asGroupEvent(event).getName();
+    } else {
+      console.error(
+        'Dialog was opened for an unsupported event type: ' + eventType
+      );
+    }
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Flow will not let you make risky things: it has seen that you declared text, textColors, backgroundColors but have not set them a value there. Always make sure that whatever the "branch" of the if/else is being used, your variables are initialised.

This is what I explained in a previous comment. Note that you could also set the variables in the "else" (`text = ''; textColor = { ... }`) but I prefer to initialise your variables as soon as possible.

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js (hunk 18 above)
Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Also you might want to remove the "a", because you're not setting it anywhere (and Flow will warn you about not setting it if it's not optional).

File: newIDE/app/src/EventsSheet/index.js [hunk 25]
```diff
@@ -1305,6 +1347,20 @@ export default class EventsSheet extends React.Component<Props, State> {
...
                           />
                         )}
+                        {this.state.textEditedEvent && (
+                          <EventTextDialog
+                            event={this.state.textEditedEvent}
+                            onApply={() => {
+                              this.closeEventTextDialog();
+                              this._saveChangesToHistory();
+                            }}
+                            onClose={() =>
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/index.js]: `onClose={this.closeEventTextDialog}`

Comment (4ian) [reply newIDE/app/src/EventsSheet/index.js]: Which you can read as English: "When [the user wants to] Close, then close the event text dialog" :)

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js (hunk 18 above)
Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Actually, as you've asked:

> And i can't directly use type RGBColor ?

I guess you refer to the one in ColorPicker. You can use it if you:
* Add the "export" keyword before "type RGBColor" in ColorPicker
* Then import it: `import ColorPicker, { type RGBColor } from '../../UI/ColorField/ColorPicker';`

I recommend that you do this :) Better have a single place where a type is defined. This type has an extra optional `a` property, but that's fine!

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js (hunk 20 above)
Comment (Bouh) [reply newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: About parseInt(), we can ignore it here ?
backgroundColor is define with getBackgroundColorXXX() and this return always a number. And ColorPicker return number too.

Unless it is necessary to do this check because the values of the.json file can be edited and falsified.

I'm wondering when it's judicious to check the type and format of the data when the methods used do already a checking.
Any way I keep it, it dosen't cost much.

Comment (4ian) [reply newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Wait a second you're right we should always have number (if we have strings there, it's a big problem somewhere).
So you can remove all the parseInt entirely: `gd.asCommentEvent(event).setTextColor(textColor.r, textColor.g, textColor.b)` :)

> Unless it is necessary to do this check because the values of the.json file can be edited and falsified.

No you can remove these parseInt, even if you go and replace them by a string in the JSON, it's being read by libGD.js which will probably read it as 0 or something like this. 
In either case, methods from libGD.js being in C++, they are guaranteed to always return a number if they declare to return a float/double/int in `Bindings.idl`. (and they also expect to receive arguments with the exact type declared in `Bindings.idl`, so if for some reason a string is sent to a function expecting a number, very bad things can happen).

> I'm wondering when it's judicious to check the type and format of the data when the methods used do already a checking.

The idea with Flow is to get the guarantee of the type and format of data. A bit more painful to write at first, but it's actually leading to cleaner code because you don't need extra check/extra conversion/extra handling of undefined values :)

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+  marginRight: 10,
+};
+
+const styles = {
+  sizeTextField: {
+    width: 90,
+    ...toolbarItemStyle,
+  },
+  toolbarItem: toolbarItemStyle,
+  checkbox: toolbarItemStyle,
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Seems unused

Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Seems unused

File: newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+const toolbarItemStyle = {
+  marginRight: 10,
+};
+
+const styles = {
+  sizeTextField: {
+    width: 90,
+    ...toolbarItemStyle,
+  },
+  toolbarItem: toolbarItemStyle,
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/InstructionEditor/EventTextDialog.js]: Seems unused

File: newIDE/app/src/EventsSheet/index.js [hunk 28]
```diff
@@ -1105,85 +1126,106 @@ export default class EventsSheet extends React.Component<Props, State> {
...
+                                click: () => this.moveEventsIntoNewGroup(),
+                              },
+                              {
+                                label: 'Analyze Objects Used in this Event',
+                                click: this._openEventsContextAnalyzer,
+                              },
+                            ];
+
+                            const edition = {
+                              label: 'Edit',
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/index.js]: Do you think we should be more precise "Edit Text and Color(s)"?

File: newIDE/app/src/EventsSheet/index.js [hunk 29]
```diff
@@ -1105,85 +1126,106 @@ export default class EventsSheet extends React.Component<Props, State> {
...
+
+                            getSelectedEvents(this.state.selection).forEach(
+                              event => {
+                                if (
+                                  event.getType() ===
+                                    'BuiltinCommonInstructions::Comment' ||
+                                  event.getType() ===
+                                    'BuiltinCommonInstructions::Group'
+                                ) {
+                                  contextList.unshift(edition);
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/index.js]: Rather than unshift-ing, let's try to keep a single array (so that it's more readable - I prefer to avoid "complex logic" in the render of components) (and also, this `render` is super long already).
Use `visible: ` set to true or false (next to label, and click) when defining the menu item.

Also careful about the forEach and unshift here, if somehow you have multiple comment or group selected, you'll have an "Edit" menu item inserted for each of them - which is probably not what you want ;)

Instead make a function for it to pass to `visible`:

`visible: filterEditableWithEventTextDialog(getSelectedEvents(this.state.selection)).length > 0`

and make `filterEditableWithEventTextDialog` a function:

```
const filterEditableWithEventTextDialog = (events: Array<gdBaseEvent>) => {
  return events.filter(event => 
    ['BuiltinCommonInstructions::Group', 'BuiltinCommonInstructions::Comment'].includes(event.getType())
  );
}
```

(not tested but you get the idea).
Last thing, where to put this function? It contains knowledge related to `EventTextDialog` so `EventTextDialog.js` file is a good place. Put it here, export it (`export const canBeEditedWithEventTextDialog = ...`).

And finally import it in EventsSheet.

This will allow to have just an array for the menu items in EventsSheet, and the "complex logic" colocated with the rest of the source code of EventTextDialog. It's better because now you only have a single file that contains all the rules telling which events are supported by EventTextDialog - rather than two (so that you have less change to forget to update one if you add a new event in the future).

File: newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css [hunk 30]
```diff
@@ -26,7 +26,11 @@
  */
 .gd-events-sheet-dark-theme .rst__moveHandle {
   background: #494949 !important;
-  border: 1px solid #494949 !important;
+  border: none;
```
Comment (4ian) [inline newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css]: Why the change the border again? Forgot what's the purpose

File: newIDE/app/src/EventsSheet/index.js [hunk 31]
```diff
@@ -362,6 +367,22 @@ export default class EventsSheet extends React.Component<Props, State> {
     return newEvents;
   };
 
+  openEventTextDialog = () => {
+    const selectedEvents = getSelectedEvents(this.state.selection);
+    if (!selectedEvents.length) return;
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/index.js]: There is a risk the last select event is not a comment/group, so better filter them (see `filterEditableWithEventTextDialog` in a next comment):

```
const editableEvents = filterEditableWithEventTextDialog(getSelectedEvents(this.state.selection));
if (!editableEvents.length) return;

const event = editableEvents[editableEvents.length - 1];  // Get the last selected event.
```

File: newIDE/app/src/EventsSheet/index.js (hunk 28 above)
Comment (Bouh) [reply newIDE/app/src/EventsSheet/index.js]: No, the sub-menus must be concise.

File: newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css (hunk 30 above)
Comment (Bouh) [reply newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css]: It's a remnant of when I split the PRs.

But i see that on darktheme, so i've deleted the border. and it need `!important;` css rule.
image

btw it's alredy done in #1309 , not yet merged :/

File: newIDE/app/src/EventsSheet/index.js [hunk 32]
```diff
@@ -1105,85 +1130,96 @@ export default class EventsSheet extends React.Component<Props, State> {
...
-                              click: () => this.toggleDisabled(),
-                              enabled: this._selectionCanToggleDisabled(),
-                            },
-                            { type: 'separator' },
-                            {
-                              label: 'Add New Event Below',
-                              click: () =>
-                                this.addNewEvent(
-                                  'BuiltinCommonInstructions::Standard'
+                          buildMenuTemplate={() => {
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/index.js]: Thanks for the changes :) Now you should be able to go back to a function directly returning an array: 
```
() => [
   ...
```

Slightly shorter/less indentation :)

File: newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css (hunk 30 above)
Comment (4ian) [reply newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css]: Alright I see, this is because of this rule: 

https://github.com/4ian/GDevelop/blob/9763d6523c797b6d721b25d444156a16d89de28d/newIDE/app/src/EventsSheet/EventsTree/style.css#L12-L15

Background color and border are set here. I think we should instead:
* Remove the border in this common rule.
* Probably remove also the background in this common rule, so that we can (probably) remove the !important that is already there on the background in the dark/default theme css files.

Comment (4ian) [reply newIDE/app/src/UI/Theme/DarkTheme/EventsSheet.css]: Note that if you want to do this in the other PR, you can remove here the changed lines.

File: newIDE/app/src/EventsSheet/index.js (hunk 32 above)
Comment (Bouh) [reply newIDE/app/src/EventsSheet/index.js]: Done ✅

**Comment by Bouh** [comment]
When you merge this think to merge #1309 with it :)

**Comment by 4ian** [comment]
Yep good remark, let me check the other PR again :)

**Comment by 4ian** [comment]
Merged and now the next one :) Thanks for adding this dialog!
